client/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -0,0 +1,476 @@
+    private volatile FailureListener failureListener = new FailureListener();
+
+    private RestClient(CloseableHttpClient client, long maxRetryTimeout, Header[] defaultHeaders, HttpHost[] hosts) {
+        this.client = client;
+        this.maxRetryTimeout = maxRetryTimeout;
+        this.defaultHeaders = defaultHeaders;
+        setHosts(hosts);
+    }
+
+    /**
+     * Replaces the hosts that the client communicates with.
+     * @see HttpHost
+     */
+    public synchronized void setHosts(HttpHost... hosts) {
+        if (hosts == null || hosts.length == 0) {
+            throw new IllegalArgumentException("hosts must not be null nor empty");
+        }
+        Set<HttpHost> httpHosts = new HashSet<>();
+        for (HttpHost host : hosts) {
+            Objects.requireNonNull(host, "host cannot be null");
+            httpHosts.add(host);
+        }
+        this.hosts = Collections.unmodifiableSet(httpHosts);
+        this.blacklist.clear();
+    }
+
+    /**
+     * Sends a request to the elasticsearch cluster that the current client points to.
+     * Selects a host out of the provided ones in a round-robin fashion. Failing hosts are marked dead and retried after a certain
+     * amount of time (minimum 1 minute, maximum 30 minutes), depending on how many times they previously failed (the more failures,
+     * the later they will be retried). In case of failures all of the alive nodes (or dead nodes that deserve a retry) are retried
+     * till one responds or none of them does, in which case an {@link IOException} will be thrown.
+     *
+     * @param method the http method
+     * @param endpoint the path of the request (without host and port)
+     * @param params the query_string parameters
+     * @param entity the body of the request, null if not applicable
+     * @param headers the optional request headers
+     * @return the response returned by elasticsearch
+     * @throws IOException in case of a problem or the connection was aborted
+     * @throws ClientProtocolException in case of an http protocol error
+     * @throws ElasticsearchResponseException in case elasticsearch responded with a status code that indicated an error
+     */
+    public ElasticsearchResponse performRequest(String method, String endpoint, Map<String, String> params,
+                                                HttpEntity entity, Header... headers) throws IOException {
+        URI uri = buildUri(endpoint, params);
+        HttpRequestBase request = createHttpRequest(method, uri, entity);
+        setHeaders(request, headers);
+        //we apply a soft margin so that e.g. if a request took 59 seconds and timeout is set to 60 we don't do another attempt
+        long retryTimeout = Math.round(this.maxRetryTimeout / (float)100 * 98);
+        IOException lastSeenException = null;
+        long startTime = System.nanoTime();
+        Iterator<HttpHost> hostIterator = nextHost();
```

Comment:
can this be a for(HttpHost host : nextHost())` if `nextHost()` would return `Iterable`?